ledger/src/shred.rs

```diff
@@ -427,98 +554,165 @@ impl Shredder {
+                                } else {
+                                    (false, false)
+                                }
+                            };
+
+                            let shred = Shred::new_from_data(
+                                self.slot,
+                                shred_index,
+                                (self.slot - self.parent_slot) as u16,
+                                Some(shred_data),
+                                is_last_data,
+                                is_last_in_slot,
+                                self.reference_tick,
+                                self.version,
+                            );
+                            let mut p = Packet::default();
+                            shred.copy_to_packet(&mut p);
+                            packets.packets.push(p);
+                        }
+                        packets.packets.resize(chunk.len(), Packet::default());
+                        packets
                     })
                     .collect()
             })
         });
-        let gen_data_time = now.elapsed().as_millis();
+        let gen_data_time = now.elapsed().as_micros();
+        datapoint_debug!(
+            "shredding-stats",
+            ("slot", self.slot as i64, i64),
+            ("num_shreds", num_shreds as i64, i64),
+            ("serialzing", serialize_time as i64, i64),
+            ("gen_data", gen_data_time as i64, i64),
+        );
+
+        (data_shreds, last_shred_index)
+    }
+
+    pub fn read_slots(batches: &[Packets]) -> HashSet<u64> {
+        batches
+            .iter()
+            .flat_map(|batch| {
+                batch.packets.iter().filter_map(|packet| {
+                    let slot_start = size_of::<Signature>() + size_of::<ShredType>();
+                    let slot_end = slot_start + size_of::<u64>();
+                    trace!("slot {} {}", slot_start, slot_end,);
+                    if slot_end <= packet.meta.size {
+                        limited_deserialize(&packet.data[slot_start..slot_end]).ok()
+                    } else {
+                        None
+                    }
+                })
+            })
+            .collect()
+    }
+
+    pub fn entries_to_shreds(
+        &self,
+        recycler_cache: &RecyclerCache,
+        entries: &[Entry],
+        is_last_in_slot: bool,
+        next_shred_index: u32,
+    ) -> (Vec<Packets>, Vec<Packets>, u32) {
+        let (mut data_shreds, last_shred_index) = self.entries_to_unsigned_data_shreds(
+            recycler_cache,
+            entries,
+            is_last_in_slot,
+            next_shred_index,
+        );
+
+        let now = Instant::now();
+        let slots = Self::read_slots(&data_shreds);
+        let mut pubkeys: HashMap<u64, [u8; 32]> = slots
+            .iter()
+            .map(|s| (*s, self.keypair.pubkey().to_bytes()))
+            .collect();
+        pubkeys.insert(std::u64::MAX, [0u8; 32]);
```

Comment:
Why is this needed?

Reply:
The memory that we send to the GPU is flat, so we don't to reorder or skip any of the packets.  if there is an error in the data it points to MAX/0..., which should fail the signature.